transformer_engine/paddle/csrc/common.h

```diff
@@ -136,10 +142,45 @@ inline DType Int2NvteDType(int64_t dtype) {
     }
 }
 
+// CUDA Utils
+class cudaDevicePropertiesManager {
+ public:
+    static cudaDevicePropertiesManager &Instance() {
+        static thread_local cudaDevicePropertiesManager instance;
+        return instance;
+    }
+
+    int GetMultiProcessorCount() {
+        if (!prop_queried_) {
+            int device_id;
+            NVTE_CHECK_CUDA(cudaGetDevice(&device_id));
+            cudaGetDeviceProperties(&prop_, device_id);
+            prop_queried_ = true;
+        }
+        return prop_.multiProcessorCount;
+    }
+
+    int GetMajor() {
+        if (!prop_queried_) {
+            int device_id;
+            NVTE_CHECK_CUDA(cudaGetDevice(&device_id));
+            cudaGetDeviceProperties(&prop_, device_id);
+            prop_queried_ = true;
+        }
+        return prop_.major;
+    }
+
+ private:
+    bool prop_queried_ = false;
+    cudaDeviceProp prop_;
```

Comment:
This assumes one GPU per process. Is this a general assumption in Paddle? If not, it would be safer to cache a separate `cudaDeviceProp` for each GPU.

Reply:
> This assumes one GPU per process. Is this a general assumption in Paddle? If not, it would be safer to cache a separate `cudaDeviceProp` for each GPU.

Sorry for the late reply due to OOTO. The assumption is always true in Paddle.